plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/RetryingJdbcClient.java

```diff
@@ -0,0 +1,527 @@
+            throws SQLException
+    {
+        // no retrying as it could be not idempotent operation (connection could be not reusable after the first failure)
+        return delegate.buildSql(session, connection, split, table, columns);
+    }
+
+    @Override
+    public CallableStatement buildProcedure(ConnectorSession session, Connection connection, JdbcSplit split, JdbcProcedureHandle procedureHandle)
+            throws SQLException
+    {
+        // no retrying as it could be not idempotent operation (connection could be not reusable after the first failure)
+        return delegate.buildProcedure(session, connection, split, procedureHandle);
+    }
+
+    @Override
+    public Optional<PreparedQuery> implementJoin(ConnectorSession session, JoinType joinType, PreparedQuery leftSource, Map<JdbcColumnHandle, String> leftProjections, PreparedQuery rightSource, Map<JdbcColumnHandle, String> rightProjections, List<ParameterizedExpression> joinConditions, JoinStatistics statistics)
+    {
+        // there should be no remote database interaction
+        return delegate.implementJoin(session, joinType, leftSource, leftProjections, rightSource, rightProjections, joinConditions, statistics);
+    }
+
+    @Override
+    public Optional<PreparedQuery> legacyImplementJoin(ConnectorSession session, JoinType joinType, PreparedQuery leftSource, PreparedQuery rightSource, List<JdbcJoinCondition> joinConditions, Map<JdbcColumnHandle, String> rightAssignments, Map<JdbcColumnHandle, String> leftAssignments, JoinStatistics statistics)
+    {
+        // there should be no remote database interaction
+        return delegate.legacyImplementJoin(session, joinType, leftSource, rightSource, joinConditions, rightAssignments, leftAssignments, statistics);
+    }
+
+    @Override
+    public boolean supportsTopN(ConnectorSession session, JdbcTableHandle handle, List<JdbcSortItem> sortOrder)
+    {
+        // there should be no remote database interaction
+        return delegate.supportsTopN(session, handle, sortOrder);
+    }
+
+    @Override
+    public boolean isTopNGuaranteed(ConnectorSession session)
+    {
+        // there should be no remote database interaction
+        return delegate.isTopNGuaranteed(session);
+    }
+
+    @Override
+    public boolean supportsLimit()
+    {
+        // there should be no remote database interaction
+        return delegate.supportsLimit();
+    }
+
+    @Override
+    public boolean isLimitGuaranteed(ConnectorSession session)
+    {
+        // there should be no remote database interaction
+        return delegate.isLimitGuaranteed(session);
+    }
+
+    @Override
+    public Optional<String> getTableComment(ResultSet resultSet)
+            throws SQLException
+    {
+        // no retrying as it could be not idempotent operation
+        return delegate.getTableComment(resultSet);
```

Comment:
why not idempotent? Do you mean when we retry we can observe different values?

Reply:
Because result set could be advanced and so we may not got proper results